File: tripin-service/src/main/java/grepp/NBE5_6_2_Team03/domain/user/code/CodeFinder.java
package grepp.NBE5_6_2_Team03.domain.user.code;

import grepp.NBE5_6_2_Team03.global.exception.BusinessException;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class CodeFinder {

    private final List<CodeProvidable> CODE_PROVIDERS = List.of(
            new PasswordCodeProvider()
    );

    public String findCodeFrom(CodeType codeType){
        return CODE_PROVIDERS.stream()
                .filter(provider -> provider.isSupport(codeType))
                .findFirst()
                .map(provider -> provider.provide(codeType))
                .orElseThrow(() -> new BusinessException("CODE_FINDER 예외"));
    }
}

File: tripin-mail/src/main/kotlin/com/grepp/mailservice/service/strategy/CodeFinder.kt
package com.grepp.mailservice.service.strategy

import org.springframework.stereotype.Component
import java.util.function.Supplier

@Component
class CodeFinder(private val providers: List<CodeProvidable>) {
    fun findCodeFrom(codeType: CodeType): String {
        return providers.find { it.isSupport(codeType) }
            ?.provide()
            ?: throw IllegalArgumentException("지원하지 않는 코드 타입입니다: $codeType")
    }
}
